### PR TITLE
Numeric const

### PR DESCRIPTION
**Pull request by stellaraccident**

### CODE REVIEW

File: include/npcomp/Dialect/Basicpy/IR/BasicpyOps.td [hunk 1]
```diff
@@ -90,53 +91,53 @@ def CompareOperationAttr : StrEnumAttr<
...
+      always include physical `IntegerType`, if only to enable progressive
+      lowering and interop with cases where the precise type is known.
+    * Python3 `float` : This is allowed to map to any legal floating point
+      type on the physical machine and is usually represented as a double (f64).
+      In MLIR, any `FloatType` is allowed, which facilitates progressive
+      lowering and interop with cases where a more precise type is known.
+    * Python3 `complex` : Maps to an MLIR `ComplexType` with a `FloatType`
+      elementType (note: in Python, complex numbers are always defined with
+      floating point components).
+    * `bool` : See `bool_constant`. In Python, bool is part of the numeric
```
Comment (silvasean) [inline include/npcomp/Dialect/Basicpy/IR/BasicpyOps.td:120]: what is the relationship to bool_constant? Does this op supercede it?

File: test/Dialect/Basicpy/ops_invalid.mlir [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// RUN: npcomp-opt -split-input-file -verify-diagnostics %s
```
Comment (silvasean) [inline test/Dialect/Basicpy/ops_invalid.mlir:1]: nit: ops-invalid.mlir ? (don't know which way the wavefunction has collapsed in this part of the repo)

File: include/npcomp/Dialect/Basicpy/IR/BasicpyOps.td [hunk 3]
```diff
@@ -230,6 +235,115 @@ def Basicpy_BytesConstantOp : Basicpy_Op<"bytes_constant", [
...
+  );
+  let assemblyFormat = "$value attr-dict";
+  let hasFolder = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// Casting and coercion operations
+//===----------------------------------------------------------------------===//
+
+def Basicpy_AsPredicateValueOp : Basicpy_Op<"as_predicate_value",
```
Comment (silvasean) [inline include/npcomp/Dialect/Basicpy/IR/BasicpyOps.td:274]: can we just call this `as_i1`? I don't see any clarity advantage to the current name.

**Review by silvasean — APPROVED** [review]

File: include/npcomp/Dialect/Basicpy/IR/BasicpyOps.td (hunk 1 above)
Comment (stellaraccident) [reply include/npcomp/Dialect/Basicpy/IR/BasicpyOps.td:120]: Commented.

**Comment by stellaraccident** [comment]
FYI - Added some folding tests, a materialization hook and asm format odds and ends that uncovered.

**Pull request merged**